src/infra/core/tests/tests/test_compact_service_impl.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::assert_matches::assert_matches;
+use std::sync::Arc;
+
+use dill::Component;
+use domain::compact_service::{CompactError, CompactService, NullCompactionMultiListener};
+use event_bus::EventBus;
+use futures::TryStreamExt;
+use kamu::domain::*;
+use kamu::testing::{DatasetTestHelper, MetadataFactory};
+use kamu::*;
+use kamu_core::{auth, CurrentAccountSubject};
+use opendatafabric::*;
+const FILE_DATA_ARRAY_SIZE: usize = 32;
+
+#[tokio::test]
+async fn test_dataset_compact() {
+    let tempdir = tempfile::tempdir().unwrap();
+    let compact_tempdir = tempfile::tempdir().unwrap();
+
+    let root_dataset_alias = DatasetAlias::new(None, DatasetName::new_unchecked("foo"));
+    let derive_dataset_alias = DatasetAlias::new(None, DatasetName::new_unchecked("derive-foo"));
+
+    let catalog = dill::CatalogBuilder::new()
+        .add::<EventBus>()
+        .add::<DependencyGraphServiceInMemory>()
+        .add_value(CurrentAccountSubject::new_test())
+        .add::<auth::AlwaysHappyDatasetActionAuthorizer>()
+        .add_builder(
+            DatasetRepositoryLocalFs::builder()
+                .with_root(tempdir.path().join("datasets"))
+                .with_multi_tenant(false),
+        )
+        .bind::<dyn DatasetRepository, DatasetRepositoryLocalFs>()
+        .add::<CompactServiceImpl>()
+        .build();
```

Comment:
Try separating catalog and other initialization code from the main test logic.
See a lot of examples of "harness" structs, where we prepare the initial state and all the plumbing, and keep test very readable at high level of abstraction.